File: backend/parser/graph_to_mermaid.py
"""
graph_to_mermaid.py — Convert graph dict to Mermaid diagram strings.
Generates: flowchart, call graph, class diagram.
"""

import re


def _safe_id(s):
    return re.sub(r"[^a-zA-Z0-9_]", "_", str(s))


def _quoted(label):
    return '"' + str(label).replace('"', "'") + '"'


# ─────────────────────────────────────────────
# FLOWCHART


def graph_to_mermaid(graph):

    return _to_flowchart(graph)


def _to_flowchart(graph):

    nodes = graph.get("nodes", {})
    edges = graph.get("edges", [])

    lines = ["flowchart TD"]

    # ───────── Nodes

    for nid, data in nodes.items():

        label = data.get("label", nid)
        ntype = data.get("type", "function")

        safe = _safe_id(nid)
        q = _quoted(label)

        if ntype == "class":
            lines.append(f"  {safe}[/{q}/]")

        elif ntype in ("entry", "exit"):
            lines.append(f"  {safe}(({q}))")

        elif ntype == "async_function":
            lines.append(f"  {safe}>{q}]")

        elif ntype == "control":
            lines.append(f"  {safe}{{{q}}}")

        elif ntype == "call":
            lines.append(f"  {safe}([{q}])")

        else:
            lines.append(f"  {safe}[{q}]")

    # ───────── Edges

    for edge in edges:

        if isinstance(edge, dict):
            src = _safe_id(edge.get("from"))
            dst = _safe_id(edge.get("to"))
            label = edge.get("label", "")

        else:
            src = _safe_id(edge[0])
            dst = _safe_id(edge[1])
            label = edge[2] if len(edge) > 2 else ""

        if label:
            lines.append(f"  {src} -->|{label}| {dst}")
        else:
            lines.append(f"  {src} --> {dst}")

    return "\n".join(lines)


# ─────────────────────────────────────────────
# CLASS DIAGRAM


def to_class_diagram(ir):

    lines = ["classDiagram"]

    classes = ir.get("classes", [])

    for cls in classes:

        name = cls.get("name")

        lines.append(f"  class {name} {{")

        for attr in cls.get("attributes", [])[:6]:
            lines.append(f"    +{attr}")

        for method in cls.get("methods", [])[:8]:
            lines.append(f"    +{method}()")

        lines.append("  }")

    # inheritance

    for cls in classes:

        for base in cls.get("bases", []):

            base_clean = base.split("[")[0].split("(")[0]

            for c in classes:
                if c.get("name") == base_clean:
                    lines.append(f"  {base_clean} <|-- {cls['name']}")

    if len(lines) == 1:
        return ""

    return "\n".join(lines)


# ─────────────────────────────────────────────
# CALL GRAPH


def to_call_graph(ir):

    funcs = ir.get("functions", [])

    if not funcs:
        return ""

    known = {f["name"] for f in funcs}

    lines = ["flowchart LR"]

    seen = set()

    for func in funcs:

        fid = _safe_id(func["name"])

        lines.append(f"  {fid}[{_quoted(func['name'])}]")

        for callee in func.get("calls", []):

            if callee in known and callee != func["name"]:

                cid = _safe_id(callee)

                edge = (fid, cid)

                if edge not in seen:
                    lines.append(f"  {fid} --> {cid}")
                    seen.add(edge)

    return "\n".join(lines)


def to_architecture(ir):
    """Generates a high-level system diagram showing module/class interactions."""
    lines = ["graph TD"]
    classes = ir.get("classes", [])
    if not classes:
        return "graph TD\n  UnknownModule[Unknown Module]"

    for c in classes:
        name = c.get("name")
        cid = _safe_id(name)
        lines.append(f"  {cid}[{_quoted(name)}]")
        for base in c.get("bases", []):
            bid = _safe_id(base)
            lines.append(f"  {bid} -.-> {cid}")
    return "\n".join(lines)


def to_sequence(ir):
    """Generates a sequence diagram based on function call order in the IR."""
    funcs = ir.get("functions", [])
    if not funcs:
        return "sequenceDiagram\n  User->>+System: call\n  System-->>-User: return"

    lines = ["sequenceDiagram", "  actor User"]
    for func in funcs[:4]:
        name = func["name"]
        fid = _safe_id(name)
        lines.append(f"  participant {fid} as {name}")
        lines.append(f"  User->>+ {fid}: call")
        for call in func.get("calls", [])[:2]:
            cid = _safe_id(call)
            lines.append(f"  {fid}->> {cid}: invoke")
        lines.append(f"  {fid}-->>- User: return")
    return "\n".join(lines)

File: backend-js/parser/graph_to_mermaid.js
/*
graphToMermaid.js — Convert graph dict to Mermaid diagram strings
Generates:
- flowchart
- class diagram
- call graph
*/

function safeId(s){

    return s.replace(/[^a-zA-Z0-9_]/g,"_")

}


function quoted(label){

    return `"${label.replace(/"/g,"'")}"`

}


// ─────────────────────────────────────────────
// Primary export

export function graphToMermaid(graph){

    return toFlowchart(graph)

}


// ─────────────────────────────────────────────
// Flowchart

function toFlowchart(graph){

    const nodes = graph.nodes || {}
    const edges = graph.edges || []

    const lines = ["graph TD"]


    // ───── Nodes ─────

    Object.entries(nodes).forEach(([nid,data])=>{

        const label = data.label || nid

        const type = data.type || "function"

        const safe = safeId(nid)

        const q = quoted(label)


        if(type==="class"){

            lines.push(`  ${safe}[/${q}/]`)

        }
        else if(type==="entry" || type==="exit"){

            lines.push(`  ${safe}((${q}))`)

        }
        else if(type==="async_function"){

            lines.push(`  ${safe}>${q}]`)

        }
        else{

            lines.push(`  ${safe}[${q}]`)

        }

    })


    // ───── Edges ─────

    edges.forEach(edge=>{

        const src = safeId(edge.from || edge[0])

        const dst = safeId(edge.to || edge[1])

        const label = edge.label || edge[2] || ""

        if(label){

            lines.push(`  ${src} -->|${label}| ${dst}`)

        }else{

            lines.push(`  ${src} --> ${dst}`)

        }

    })


    return lines.join("\n")

}


// ─────────────────────────────────────────────
// Class Diagram

export function toClassDiagram(ir){

    const lines = ["classDiagram"]

    const classes = ir.classes || []


    classes.forEach(cls=>{

        const name = cls.name

        lines.push(`  class ${name} {`)


        (cls.attributes || []).slice(0,6).forEach(attr=>{

            lines.push(`    +${attr}`)

        })


        (cls.methods || []).slice(0,8).forEach(method=>{

            lines.push(`    +${method}()`)

        })


        lines.push("  }")

    })


    // inheritance

    classes.forEach(cls=>{

        (cls.bases || []).forEach(base=>{

            const clean = base.split("[")[0].split("(")[0]

            const exists = classes.some(c=>c.name===clean)

            if(exists){

                lines.push(`  ${clean} <|-- ${cls.name}`)

            }

        })

    })


    return lines.length>1 ? lines.join("\n") : ""

}


// ─────────────────────────────────────────────
// Call Graph

export function toCallGraph(ir){

    const funcs = ir.functions || []

    if(!funcs.length) return ""


    const known = new Set(
        funcs.map(f=>f.name)
    )

    const lines = ["graph LR"]

    const seen = new Set()


    funcs.forEach(func=>{

        const fid = safeId(func.name)

        lines.push(`  ${fid}[${quoted(func.name)}]`)


        (func.calls || []).forEach(callee=>{

            if(
                known.has(callee) &&
                callee !== func.name
            ){

                const cid = safeId(callee)

                const edge = `${fid}-${cid}`

                if(!seen.has(edge)){

                    lines.push(`  ${fid} --> ${cid}`)

                    seen.add(edge)

                }

            }

        })

    })


    return lines.join("\n")

}